File: lecture/dingcorithm/2st_week/02_04_add_node_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self, value):
        self.head = Node(value)

    def append(self, value):
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = Node(value)

    def print_all(self):
        cur = self.head
        while cur is not None:
            print(cur.data)
            cur = cur.next

    def get_node(self, index):
        cur_index = 0
        cur = self.head
        while cur_index != index:
            cur = cur.next
            cur_index += 1
        return cur
    # index 위치에 value를 가진 Node를 추가해주세요
    def add_node(self, index, value):
        new_node = Node(value)
        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return
        prev_node = self.get_node(index - 1)
        next_node = prev_node.next
        prev_node.next = new_node
        new_node.next = next_node


    # index 번쨰 노드를 제거해주세요
    def delete_node(self, index):
        if index == 0:
            self.head = self.head.next
            return
        node = self.get_node(index-1)
        node.next = node.next.next



linked_list = LinkedList(5)
linked_list.append(12)
print(linked_list.get_node(0).data) # -> 5를 들고 있는 노드를 반환해야 합니다!

File: study/dingcorithm/2st_week/02_04_add_node_list.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor(data) {
    this.head = new Node(data);
  }

  append(data) {
    const newNode = new Node(data);

    // 빈 리스트일 경우도 고려필요
    if (!this.head) {
      this.head = newNode;
      return;
    }

    let cur = this.head;

    while (cur.next !== null) {
      cur = cur.next;
    }

    cur.next = newNode;
  }

  printAll() {
    let cur = this.head;

    while (cur !== null) {
      console.log(cur.data);
      cur = cur.next;
    }
  }

  getNode(index) {
    let cur = this.head;
    let count = 0;

    while (cur !== null && count < index) {
      cur = cur.next;
      count++;
    }

    return cur;
  }

  addNode(index, data) {
    const newNode = new Node(data);

    if (index === 0) {
      newNode.next = this.head;
      this.head = newNode;
      return;
    }

    const prevNode = this.getNode(index - 1);
    if (!prevNode) return;

    newNode.next = prevNode.next;
    prevNode.next = newNode;
  }

  deleteNode(index) {
    // index === 0 일 경우를 예외 처리한다
    const targetNode = this.getNode(index);
    if (index === 0) {
      this.head = targetNode.next;
      return;
    }

    // 이전 노드를 찾는다
    const prevNode = this.getNode(index - 1);
    // 삭제할 다음 노드와 연결한다
    const nextNode = targetNode.next;
    prevNode.next = nextNode;
  }
}

const sampleNode = new Node(3);

const list = new LinkedList(3);
list.append(6);
list.append(7);
list.append(12);

list.addNode(1, 6);
list.addNode(0, 7);
list.printAll();

console.log("---------------");
list.deleteNode(4);
list.deleteNode(0);
list.printAll();
